Allow older versions in the *.num files

mattcaswell opened the pull request:
In 1.1.0 we only allowed a strictly increasing version number in the *.num
files, i.e. you could never introduce a symbol at the end of the *.num file
with a lower version number than the one preceding it. This made sense for
1.1.0. However in master we may be introducing symbols for backport to
1.1.0. Therefore it is ok in master to have a symbol for version 1.1.0c
coming after a symbol for version 1.1.1.

This commit fixes the check in mkdef.pl to be a bit looser to allow this.

t-j-h commented:
Should it emit a warning in case the change was inadvertent - i.e. allow being non strict but warn in case the user isn't in the context you are allowing?

mattcaswell commented:
Could do...but the reality is any warning won't be seen as this script gets called in the middle of the build run.

levitte commented:
Why would 1.1.0c get an added function?

mattcaswell commented:
Well, for example, missing accessors. If you look in libcrypto.num in master you will see that there are already some instances of 1.1.0a symbols that we have added there and backported. It worked on that occasion because at that point we hadn't added any 1.1.1 symbols, so everything was still in version number order. We have now added 1.1.1 symbols, so we can't backport any more until this is fixed.

levitte commented:
Ah, good point!

kroeckx commented:
On Sat, Oct 15, 2016 at 04:36:24PM -0700, Richard Levitte wrote:
> Why would 1.1.0c get an added function? 

In the past we've also had to add new functions for a security
issue.

mattcaswell commented:
Pushed.

kaduk commented:
Should this also get merged to 1.1.0, in case a function needs to be added for a security patch that will also be backported to 1.0.2?

mattcaswell commented:
> Should this also get merged to 1.1.0, in case a function needs to be added for a security patch that will also be backported to 1.0.2?

1.0.2 doesn't have version numbers so it shouldn't matter. Even if it did 1.0.2 and 1.1.0 are not binary compatible so the same symbols would have different version numbers anyway.

kaduk commented:
Okay, so in that case the newly added symbols would just get the corresponding 1.1.0x versions and everything would still work, got it.
